### ISSUE TITLE
[Q&A] Please click on the object that appears only once.

### ISSUE THREAD

**Issue by PieceOfGood** [QIN2DIM/hcaptcha-challenger#806]
### Prompt[en]

Please click on the thumbnail of something that can be eaten

### request_type

image_label_area_select: point (fox ex. Please click on the X)

### Sitekey

https://accounts.hcaptcha.com/demo?sitekey=

### Sitelink

https://accounts.hcaptcha.com/demo?sitekey=a5f74b19-9e45-40e0-b45d-47ff91b7a6c2

### Screenshot of the challenge

![eat](https://github.com/QIN2DIM/hcaptcha-challenger/assets/31653906/1201d966-d867-4ad8-b137-928ac31bc127)

**Comment by QIN2DIM** [comment]
dup

**Comment by PieceOfGood** [comment]
Where?
Is there a solution described somewhere? I haven't found a way.

**Comment by QIN2DIM** [comment]
https://[github.com/QIN2DIM/hcaptcha-challenger/issues/725](https://github.com/QIN2DIM/hcaptcha-challenger/issues/725)

**Comment by PieceOfGood** [comment]
I'm sorry. I found this later. But I can't find a way that solves this problem.
For example, next image:
![i02](https://github.com/QIN2DIM/hcaptcha-challenger/assets/31653906/ad0a096e-cfe5-4053-9b13-99ea9c67db24)
And prompt: `Please click on the object that appears only once.`
I try:
```python
from pathlib import Path
import hcaptcha_challenger as solver
from hcaptcha_challenger import AreaSelector

solver.install(upgrade=True, flush_yolo=[solver.DEFAULT_KEYPOINT_MODEL])
IMAGE = Path(__file__).parent.joinpath("image.png").read_bytes()

def main() -> None:
    prompt = "Please click on the object that appears only once."
    tool = AreaSelector()
    results = tool.execute(prompt, [IMAGE], shape_type="point")
    print("Results:", results)
```
This prints:
![image](https://github.com/QIN2DIM/hcaptcha-challenger/assets/31653906/a6e4f823-6b25-49b0-bef7-2aae0281d612)

Can you point out what the error is?

**Comment by QIN2DIM** [comment]
Looking at the source code, this is a solvable problem.

I just didn't make it into an example. the documentation for it I didn't have time to write.

**Comment by PieceOfGood** [comment]
Thank you for your work and attention to my problem. I hope you find time to supplement the examples with solutions for this type of problem.

**Comment by joewongywai** [comment]
> Thank you for your work and attention to my problem. I hope you find time to supplement the examples with solutions for this type of problem.

Hi. Can you tell me how to solve this kind of 'appears only once' problem?

**Comment by PieceOfGood** [comment]
> > Thank you for your work and attention to my problem. I hope you find time to supplement the examples with solutions for this type of problem.
> 
> Hi. Can you tell me how to solve this kind of 'appears only once' problem?

I have no idea.
The tutorial says that for all tasks there are only two components: `Classifier` and `AreaSelector`. No one solves this problem. As you can see from this conversation, the author refers to a lack of time, which does not allow him to supplement the examples with the necessary solution. Judging by the activity, this is indeed the case and there are problems with much higher priority. I suggest a wait, since the author is aware of this situation.

**Comment by joewongywai** [comment]
> > > Thank you for your work and attention to my problem. I hope you find time to supplement the examples with solutions for this type of problem.
> > 
> > 
> > Hi. Can you tell me how to solve this kind of 'appears only once' problem?
> 
> I have no idea. The tutorial says that for all tasks there are only two components: `Classifier` and `AreaSelector`. No one solves this problem. As you can see from this conversation, the author refers to a lack of time, which does not allow him to supplement the examples with the necessary solution. Judging by the activity, this is indeed the case and there are problems with much higher priority. I suggest a wait, since the author is aware of this situation.

Hi. I found the codes to solve the 'appears only once' problems. 
Check the pull request #764 especially the control.py and the cv_toolkit.py

**Comment by PieceOfGood** [comment]
> > > > Thank you for your work and attention to my problem. I hope you find time to supplement the examples with solutions for this type of problem.
> > > 
> > > 
> > > Hi. Can you tell me how to solve this kind of 'appears only once' problem?
> > 
> > 
> > I have no idea. The tutorial says that for all tasks there are only two components: `Classifier` and `AreaSelector`. No one solves this problem. As you can see from this conversation, the author refers to a lack of time, which does not allow him to supplement the examples with the necessary solution. Judging by the activity, this is indeed the case and there are problems with much higher priority. I suggest a wait, since the author is aware of this situation.
> 
> Hi. I found the codes to solve the 'appears only once' problems. Check the pull request #764 especially the control.py and the cv_toolkit.py

You are probably talking about [this code](https://github.com/QIN2DIM/hcaptcha-challenger/blob/d97e1e62303293957897d925a3a216384809afe5/hcaptcha_challenger/agents/playwright/control.py#L429). Perhaps it is possible to make a solution based on this code, but it would be better to have it from the functionality of this library, since it is positioned on this and one of the implementations of the solution already exists.

Thanks for this tip, but I still hope for the professional help of the author, so as not to reinvent my wheel when someone already knows how to do it in the best way.

**Comment by QIN2DIM** [comment]
@PieceOfGood Hi,
This method consists of two steps: 1. object detection 2. clustering comparison

Specifically.

1. I detect the inserted foreground targets in the image by two schemes: YOLOv8(COCO pre-train) Model and Hough circle detection.

2. I segment these regions as circles of equal length and radius 32. 3.

3. use spectral clustering to find the circular object with the largest deviation

4. return the coordinates of the center of the circle of this target.


But there are a couple of big problems with this process:

1. Hough circle detection is not precise enough (I could detect all circles perfectly using a fine-tuned YOLO model, but I haven't had time to work on that lately), and on some examples it is not possible to find all circles, or there are exactly missing "circles that appear only once". As a result, the clustering results are affected
2. The clustering method requires fine-tuning of the parameters, or pre-processing of the images.

I had planned to use a cross-modal scheme like CLIP or ViLT to do zero-shot or train a miniature, fine-tuned YOLOv8-seg model, but as mentioned above, this may take a bit of time.

In addition, I'm planning to do mobile deployable zero-shot image classification/object detection solutions. If you have any good ideas you can share them with me here.

**Comment by PieceOfGood** [comment]
Thank you for such a detailed answer.
Unfortunately, my knowledge of using neural networks is far from even initial, and in my search for solutions I expected to find something that works on the principle of: “put 'A' inside 'B' and throw towards the moon with acceleration 'N'.” Examples from the guide to this tool demonstrate approximately this diagram, but when I tried to use it, the first task was one that was not among them.

Considering how relevant what you are doing is in our time, I would really like to help you, but as it turns out, I myself need your help. And apparently there are a lot of people in this situation and there will only be more of them.

I understand the situation to the extent accessible to me and I ask you to forgive me for my reluctance to fight my ignorance, but this is a separate discipline that requires time to study, which I also don’t have, which is essentially the reason for my search and the circle closes at this point .

I hope you have time or a worthy assistant to overcome this limitation.

**Comment by QIN2DIM** [comment]
Actually, I patch this question yesterday, and now its accuracy should be close to 100%.😅

**Comment by joewongywai** [comment]
> Actually, I patch this question yesterday, and now its accuracy should be close to 100%.😅

I appreciate that you let us know your commits by replying questions in the discussion.
Please keep this habit if possible.
Using the structural_similarity function is quite smart and doesn't require spectral clustering you used in the previous commit.
I also found from your blog post that your spectral  clustering approach in the previous commit was inspired by GPT. I wonder if you could ask GPT in future to also find solutions from the existing python packages (e.g. the structural_similarity from skimage package). I hope this will speed up the updates of this repository so that we will have a hcaptcha solver with high accuracy. :)

**Comment by PieceOfGood** [comment]
I'm sorry, do you have an example of how this works, such as the code demos from your guide.

> Решение, которое я раньше использовал, возникло в процессе общения с GPT, и я попробовал использовать инфраструктуру llm в различных ссылках.
> 
> Другими словами, я надеюсь, что в будущем смогу найти способ разгадать все капчи в мире с помощью LLM.
> 
> Но в настоящее время я очень рад тому, что GPT может достичь такого уровня в однотекстовом режиме, но метод точности, который он дает, на самом деле недостаточно высок.
> 
> Поэтому я использовал@beiyuouo Модель сегмента YOLOv8 + метод расчета матрицы сходства для решения этой задачи.

I'm sorry, do you have an example of how this works, such as the code demos [from your guide](https://github.com/QIN2DIM/hcaptcha-challenger/wiki/Get-started-%E2%80%90-Library).
